src/item/enchantment/Enchantment.php

```diff
@@ -92,5 +123,34 @@ public function getMaxLevel() : int{
 		return $this->maxLevel;
 	}
 
-	//TODO: methods for min/max XP cost bounds based on enchantment level (not needed yet - enchanting is client-side)
+	/**
+	 * Returns whether this enchantment can be applied to the item along with the given enchantment.
+	 */
+	public function isCompatibleWith(Enchantment $other) : bool{
+		return IncompatibleEnchantmentRegistry::getInstance()->areCompatible($this, $other);
+	}
+
+	/**
+	 * Returns the minimum enchanting power value required for the particular level of the enchantment
+	 * to be available in an enchanting table.
+	 *
+	 * Enchanting power is a random value based on the number of bookshelves around an enchanting table
+	 * and the enchantability of the item being enchanted. It is only used when determining the available
+	 * enchantments for the enchantment options.
+	 */
+	public function getMinEnchantingPower(int $level) : int{
+		return ($this->minEnchantingPower)($level);
+	}
+
+	/**
+	 * Returns the maximum enchanting power value allowed for the particular level of the enchantment
+	 * to be available in an enchanting table.
+	 *
+	 * Enchanting power is a random value based on the number of bookshelves around an enchanting table
+	 * and the enchantability of the item being enchanted. It is only used when determining the available
+	 * enchantments for the enchantment options.
+	 */
+	public function getMaxEnchantingPower(int $level) : int{
```

Comment:
Just so I understand this. If there are too many bookshelves, or the item's enchantability is very high, the enchantment will be removed from the selection pool?

Reply:
That's about right. If the random enchantingPower value, which is based on the number of bookshelves and the item's enchantability, exceeds maxEnchantingPower for a given enchantment level, that enchantment level is removed from the selection pool.

Reply:
enchantingPower is generating in `EnchantmentHelper::createEnchantOption`, you can check it